sdk/identity/azure-identity/azure/identity/_authn_client.py

```diff
@@ -35,12 +36,13 @@ def __init__(self, auth_url, **kwargs):
         self._cache = TokenCache()
 
     def get_cached_token(self, scopes):
-        # type: (Iterable[str]) -> Optional[str]
+        # type: (Iterable[str]) -> Optional[AccessToken]
         tokens = self._cache.find(TokenCache.CredentialType.ACCESS_TOKEN, list(scopes))
         for token in tokens:
             if all((scope in token["target"] for scope in scopes)):
-                if int(token["expires_on"]) - 300 > int(time()):
-                    return token["secret"]
+                expires_on = int(token["expires_on"])
+                if expires_on - 300 > int(time()):
+                    return AccessToken(token["secret"], expires_on)
```

Comment:
If the token expires, maybe we should remove it from the cache.

Reply:
The cache stores only one value per key. When we retrieve a token that fails this expiry check, we're about to overwrite it with a new one.